File: src/main/java/net/dblsaiko/retrocomputers/common/init/Resources.java
package net.dblsaiko.retrocomputers.common.init;

import net.minecraft.resource.ResourceManager;
import net.minecraft.resource.ResourceType;
import net.minecraft.util.Identifier;
import net.minecraft.util.profiler.Profiler;
import net.fabricmc.fabric.api.resource.ResourceManagerHelper;
import net.fabricmc.fabric.api.resource.SimpleResourceReloadListener;

import java.io.IOException;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;
import java.util.stream.Collectors;

import net.dblsaiko.hctm.init.RegistryObject;
import net.dblsaiko.retrocomputers.RetroComputers;

import static net.dblsaiko.retrocomputers.RetroComputersKt.MOD_ID;

public record Resources(
    byte[] bootloader,
    byte[] charset,
    Map<Identifier, byte[]> disks
) {
    private static final Identifier ID = new Identifier(MOD_ID, "data");

    public static void register() {
        ResourceManagerHelper.get(ResourceType.SERVER_DATA).registerReloadListener(new SimpleResourceReloadListener<Resources>() {
            @Override
            public CompletableFuture<Resources> load(ResourceManager manager, Profiler profiler, Executor executor) {
                return CompletableFuture.supplyAsync(() -> {
                    byte[] bootloader = loadImage(manager, "bootldr.bin");
                    byte[] charset = loadImage(manager, "charset.bin");
                    Map<Identifier, byte[]> disks =
                        RetroComputers.INSTANCE.getItems().getSysDiskObjects()
                            .stream()
                            .collect(Collectors.toUnmodifiableMap(
                                RegistryObject::id,
                                k -> loadImage(manager, "disks/%s.img".formatted(k.get().getImage().getPath()))
                            ));
                    return new Resources(bootloader, charset, disks);
                }, executor);
            }

            @Override
            public CompletableFuture<Void> apply(Resources data, ResourceManager manager, Profiler profiler, Executor executor) {
                return CompletableFuture.runAsync(() -> {
                    RetroComputers.INSTANCE.setResources(data);
                }, executor);
            }

            @Override
            public Identifier getFabricId() {
                return ID;
            }
        });
    }

    private static byte[] loadImage(ResourceManager manager, String name) {
        try (var res = manager.getResource(new Identifier(MOD_ID, name))) {
            return res.getInputStream().readAllBytes();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/kotlin/common/init/Resources.kt
package net.dblsaiko.retrocomputers.common.init

import net.dblsaiko.retrocomputers.MOD_ID
import net.fabricmc.fabric.api.resource.IdentifiableResourceReloadListener
import net.fabricmc.fabric.api.resource.ResourceManagerHelper
import net.minecraft.resource.ResourceManager
import net.minecraft.resource.ResourceReloadListener.Synchronizer
import net.minecraft.resource.ResourceType.SERVER_DATA
import net.minecraft.util.Identifier
import net.minecraft.util.profiler.Profiler
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executor

object Resources {

  private lateinit var bootloader: ByteArray
  private lateinit var charset: ByteArray

  private lateinit var disks: Map<Identifier, ByteArray>

  fun bootloader() = bootloader.clone()

  fun charset() = charset.clone()

  fun disk(id: Identifier) = disks.getValue(id).clone()

  internal fun register() {
    ResourceManagerHelper.get(SERVER_DATA).registerReloadListener(object : IdentifiableResourceReloadListener {

      override fun reload(s: Synchronizer, rm: ResourceManager, profiler: Profiler, profiler1: Profiler, executor: Executor, executor1: Executor): CompletableFuture<Void> {
        return CompletableFuture.runAsync(Runnable {
          bootloader = rm.getResource(Identifier(MOD_ID, "bootldr.bin")).inputStream.use { it.readBytes() }
          charset = rm.getResource(Identifier(MOD_ID, "charset.bin")).inputStream.use { it.readBytes() }
          disks = Items.SYS_DISKS.associate { it.image to rm.getResource(Identifier(it.image.namespace, "disks/${it.image.path}.img")).inputStream.use { it.readBytes() } }
        }, executor).thenCompose<Void> { s.whenPrepared(null) }
      }

      override fun getFabricId(): Identifier = Identifier(MOD_ID, "data")
    })
  }

}